### ISSUE TITLE
Feature request: open files w/ name and size labels

### ISSUE THREAD

**Issue by ageis** [ncabatoff/process-exporter#75]
It would be excellent to have the capability to aggregate metrics of each processes's open FDs that are symlinks to actual files (as opposed to sockets or device nodes), via proc.FileDescriptorTargets(), and export the file name and file size in bytes. This has the potential create a lot of metrics; so perhaps it should be off by default and have a corresponding CLI flag to enable. I've been taking a stab at implementing this myself but I am new to Go and slightly confused by the way this exporter is factored together.

```
			for file := range *p.Proc.FileDescriptorTargets() {
				var filename string
				var filesize string
				fd, err := os.Open(file)
				if err != nil {
					return
				}
				defer fd.Close()
				stat, err := fd.Stat()
				if err != nil {
					return
				}
				filesize = stat.Size()
				filename = file
				ch <- prometheus.MustNewConstMetric(openFileDesc,
					prometheus.GaugeValue, float64(p.Proc.FileDescriptorsLen()), filename, filesize)
			}
```

**Comment by ncabatoff** [comment]
Hi @ageis ,

I'm pretty sure I don't want to start publishing metrics that include individual file names and file sizes.  I might be open to breaking down the count of open FDs by type, so you could see how many sockets vs files, devices, directories, pipes, etc.

What problem(s) are you trying to solve with this feature?

**Comment by ageis** [comment]
@ncabatoff I had a potential use case in a previous job, but no longer. It's your project so feel free to close if you don't feel the utility...

I still think such a feature (optional and disabled by default) could be useful for auditing, for example providing alerts when certain files are read; in addition to the ability to calculate average (and other metrics) from file sizes under certain workloads, flagging particularly large files, etc.

**Comment by ncabatoff** [comment]
It's not that I don't see the utility, it's more that there are plenty of features that would provide *some* value, and I try to judge each based on how well they fit with the primary mission of this exporter and how much benefit they provide to its users relative to how much overhead maintaining it will take.  

A breakdown of open files by type: quite useful in some cases, because it could allow you to track file descriptor leaks, socket proliferation due to badly behaved or malicious clients, etc.  It also has only a small additional incremental cost so it could be enabled by default.

Whereas the optional file size by name feature you describe would be rarely used by most existing process-exporter users and seems more suited to a more specialized tool, or perhaps just a cronjob that writes to the node_exporter text file directory.

**Comment by ageis** [comment]
Ah, I remembered something anecdotal. I ended up using [osquery_exporter](https://github.com/zwopir/osquery_exporter) to obtain the list of open files that I'd originally envisioned gathering with process-exporter, if my Go chops had been good enough.

(example stanza of osquery_exporter config):
```
  gaugevecs:
    -
      name: open_files
      help: "files opened by processes"
      query: "select distinct pid, path from process_open_files group by path;"
      valueidentifier: pid
      labelidentifier:
       - path  
```

**Comment by ncabatoff** [comment]
Very cool, thanks for letting me know about it.  If I ever fix #77 it'll probably be by wrapping that exporter.

**Comment by ncabatoff** [comment]
Closing since I don't plan to implement this.
